### PR TITLE
Check branches with a configured remote are pushed

### PR DESCRIPTION
**Pull request by evnu**
Retrieve all branches with a configured remote. For all, use git log to check if they have commits which are not present on the remote branch.

Give this a spin before merging it. It worked fine during my local tests, but maybe I missed some edge cases.

### CODE REVIEW

File: checker.py [hunk 1]
```diff
@@ -227,9 +225,27 @@ def git_checker(
...
             # WD dirty
             dirty_dirs += [gd]
-        if check_unpushed(lines):
-            # Changes unpushed
-            unpushed_dirs += [gd]
+
+    # Then, check for unpushed branches with a configured remote. This is done by using
+    # git log remote/branch..branch.
+    unpushed_branches: List[str] = []
+    for gd in itr:
```
Comment (mbforbes) [inline checker.py:221]: Do you think we can keep this under the same loop (`for gd in itr`) above? That way we won't need two progress bars to run this.

File: checker.py [hunk 2]
```diff
@@ -227,9 +225,27 @@ def git_checker(
...
+        # Retrieve all branches with a configured remote
```
Comment (mbforbes) [inline checker.py]: Now that this doesn't use the result of `git status`, what do you think about keeping this in its own function, and moving the subprocesses into their functions (i.e., `git status` is run in `check_clean()`, and `git config` is run in `check_unpushed()`? I worry this function is too big which hurts readability.

File: checker.py [hunk 3]
```diff
@@ -239,17 +255,17 @@ def git_checker(
...
             len(dirty_dirs), reportify(dirty_dirs)
         )
-    if len(dirty_dirs) > 0 and len(unpushed_dirs) > 0:
+    if len(dirty_dirs) > 0 and len(unpushed_branches) > 0:
         report += '\n\n'
-    if len(unpushed_dirs) > 0:
-        report += 'The following directories ({}) need to be pushed:\n{}'.format(
-            len(unpushed_dirs), reportify(unpushed_dirs)
+    if len(unpushed_branches) > 0:
+        report += 'The following directories+branches ({}) need to be pushed:\n{}'.format(
```
Comment (mbforbes) [inline checker.py]: If we go with the above, this could be `...directories (+branches)` to show the branches appear optionally.

File: checker.py [hunk 4]
```diff
@@ -227,9 +225,27 @@ def git_checker(
...
+        # Check each branch with a remote which is not pushed
+        for (branch, remote) in branches_with_remotes:
+            # Check which commits are on branch, but not on remote/branch
+            query = f"{remote}/{branch}..{branch}"
+            p = sp.Popen(['git', 'log', query], stdout=sp.PIPE, universal_newlines=True, cwd=gd)
+            res, ess = p.communicate()
+            # Find what's important.
+            if res != "":
+                # Changes unpushed
+                unpushed_branches += [f"{gd}, branch {branch}"]
```
Comment (mbforbes) [inline checker.py]: What do you think about omitting reporting the branch if it is `master` and it is the only one? This could keep the output simpler.

**Review by mbforbes — COMMENTED** [review]
This is a great proposal! Left some suggestions below, let me know what you think :-)

File: checker.py (hunk 1 above)
Comment (evnu) [reply checker.py:221]: With ac2b98c, I pulled the two loops into one loop.

File: checker.py (hunk 2 above)
Comment (evnu) [reply checker.py]: With ac2b98c, I pulled the actual checks into separate functions. What do you think?

File: checker.py (hunk 4 above)
Comment (evnu) [reply checker.py]: Done with d09c54c.

**Comment by mbforbes** [comment]
This looks great! Thanks for such a fast turnaround. Let me check out your branch and give it a spin before merging!

File: checker.py (hunk 2 above)
Comment (mbforbes) [reply checker.py]: LGTM! I like this style --- I always return results rather than mutate function args, but this actually saves lines of code, which is really nice.

**Comment by mbforbes** [comment]
Tried it out — LGTM!
